layout/plan.c

```diff
@@ -178,12 +237,17 @@ M0_INTERNAL void m0_layout_plan_fini(struct m0_layout_plan *plan)
 	m0_tl_teardown(pplops, &plan->lp_plops, plop) {
 		if (plop->pl_ops && plop->pl_ops->po_fini)
 			plop->pl_ops->po_fini(plop);
-		/* for each plan_get() plop_done() must be called */
+		/* For each plan_get(), plop_done() must be called. */
 		M0_ASSERT(M0_IN(plop->pl_state, (M0_LPS_INIT, M0_LPS_DONE)));
+		del_plops_relation(pldeps_tlist_head(&plop->pl_deps));
 		m0_free(plop);
 	}
 	pplops_tlist_fini(&plan->lp_plops);
 	plan->lp_op = NULL;
+
+	m0_mutex_unlock(&plan->lp_lock);
+	m0_mutex_fini(&plan->lp_lock);
```

Comment:
Similarly to _init(): if lock-unlock is useful (i.e., concurrency is possible at this point), then mutex_fini() is invalid, because other thread can hold the lock at this point.

Reply:
Agree, concurrency should not be possible at this point. But don't we want to make sure that it is so by making the assert-check for all plops state when we destroy them here in the loop atomic in respect to other plop ops?

Reply:
Documented.